workers/unity/Packages/com.improbable.gdk.playerlifecycle/Systems/PlayerHeartbeat/HandlePlayerHeartbeatResponseSystem.cs

```diff
@@ -1,101 +1,87 @@
-                All = new[]
-                {
-                    ComponentType.Create<HeartbeatData>(),
-                    ComponentType.Create<WorldCommands.DeleteEntity.CommandSender>(),
-                    ComponentType.ReadOnly<PlayerHeartbeatClient.CommandResponses.PlayerHeartbeat>(),
-                    ComponentType.ReadOnly<SpatialEntityId>(),
-                },
-                Any = Array.Empty<ComponentType>(),
-                None = Array.Empty<ComponentType>()
-            };
-
-            group = GetComponentGroup(query);
+            workerSystem = World.GetExistingManager<WorkerSystem>();
+            commandSystem = World.GetExistingManager<CommandSystem>();
+            logger = workerSystem.LogDispatcher;
         }
 
         protected override void OnUpdate()
         {
-            var heartbeatType = GetArchetypeChunkComponentType<HeartbeatData>();
-            var entityDeleterType = GetArchetypeChunkComponentType<WorldCommands.DeleteEntity.CommandSender>();
-            var responsesType =
-                GetArchetypeChunkComponentType<PlayerHeartbeatClient.CommandResponses.PlayerHeartbeat>(true);
-            var spatialIdType = GetArchetypeChunkComponentType<SpatialEntityId>(true);
+            var responses = commandSystem.GetResponses<PlayerHeartbeatClient.PlayerHeartbeat.ReceivedResponse>();
+            if (responses.Count == 0)
+            {
+                return;
+            }
 
-            var chunkArray = group.CreateArchetypeChunkArray(Allocator.TempJob);
+            var heartbeats = GetComponentDataFromEntity<HeartbeatData>();
 
-            foreach (var chunk in chunkArray)
+            // Concatenate command responses
+            foreach (var response in responses)
             {
-                var heartbeats = chunk.GetNativeArray(heartbeatType);
-                var responses = chunk.GetNativeArray(responsesType);
-                var deleteRequesters = chunk.GetNativeArray(entityDeleterType);
-                var spatialIds = chunk.GetNativeArray(spatialIdType);
-
-                for (var i = 0; i < responses.Length; i++)
+                var spatialId = response.EntityId;
+                if (!workerSystem.TryGetEntity(spatialId, out var entity) || !heartbeats.Exists(entity))
                 {
-                    var heartbeatData = heartbeats[i];
+                    continue;
+                }
 
-                    var responded = false;
-                    foreach (var response in responses[i].Responses)
-                    {
-                        if (response.StatusCode == StatusCode.Success)
-                        {
-                            responded = true;
-                            break;
-                        }
-                    }
+                var key = (spatialId, entity);
+                respondedCache.TryGetValue(key, out var responded);
+                respondedCache[key] = responded | (response.StatusCode == StatusCode.Success);
```

Comment:
Don't you lose information if you receive two heartbeat responses in the same frame of: success, failed? 

Not that this is very likely and not very problematic?

Reply:
It does, but since the goal is not to actually count heartbeats, but just ensure we keep the entity around while there is a connection, this was a simple way to ensure there are no false positives